### ISSUE TITLE
Distributions with and without go-srpm-macros

### ISSUE THREAD

**Issue by ingvagabund** [gofed/gofed#29]
Spec files should be generated on F22+ and RHEL distributions.

Assumptions:
- F22+ has go-srpm-macros in minimal buildroot (gopath, go_arches macros)
- el6, el7 has no go-srpm-macros, golang package provides gopath macro

For F22+, spec files will BuildRequires virtual compiler(go-compiler). It provides gobuild and gotest macros (among others) which simplifies %build section.
For el6, el7, spec files will BuildRequires golang as it is implicit. Once there will be need for SA on el7, spec files get updated per piece.

As it is aim to cover as much architectures and distributions, all spec files will use %gobuild macro to build golang projects. Where there is no such macro, it will get defined for golang with BuildID generated. Thus no matter where the spec file is build (architecture, with or without debug info), it will be buildable no matter of distribution.

1) Proposing the following form of ExclusiveArch and BuildRequires:

``` vim
# e.g. el6 has ppc64 arch without gcc-go, so EA tag is required
ExclusiveArch: %{?go_arches:%{go_arches}}%{!?go_arches:%{ix86} x86_64 %{arm}}
# If go_compiler is not set to 1, there is no virtual provide. Use golang instead.
BuildRequires: %{?go_compiler:compiler(go-compiler)}%{!?go_compiler:golang}
```

el7 will then use golang and build on x86_64 as well without need for go-srpm-macros

2) Proposing the following use of gobuild and gotest macros:

``` vim
%gobuild -o bin/go-md2man %{import_path}
...
%gotest %{import_path}/TEST
```

Header of each spec file that builds a golang project contains:

``` vim
%if ! 0%{?gobuild:1}
%define gobuild(o:) go build -ldflags "${LDFLAGS:-} -B 0x$(head -c20 /dev/urandom|od -An -tx1|tr -d ' \\n')" -a -v -x %{?**};
%endif

%if ! 0%{?gotest:1}
%define gotest() go test -ldflags "${LDFLAGS:-}" %{?**}
%endif
```

At the moment, gobuild macro supports only -o option as the option is common for both gcc-go and golang compilers. The aim is to support only what is really needed and minimize a change of providing option that is supported only by one compiler.

**Comment by ingvagabund** [comment]
Fedora and EPEL branches (el6, epel7) has common git repository. So I would propose to introduce go-srpm-macros to el6 as well. However, it can not be in the minimal buildroot as there is no gcc-go for el6 and most likely never will be. At the same time go-compilers package would go into f21+ and el6 and thus brought gobuild and gotest macros. So no additional definition of gobuild and gotest at the top of a spec file. As RHEL7 has different repository, spec file can differ. However, as to make packaging as simple as possible, RHEL7 spec file should be as close as possible. It is possible to introduce go-srpm-macros and go-compilers to RHEL7. At the moment, I suggest to just define gobuild/gotest for RHEL7 only at the top of each spec file and assume there is no go-srpm-macros no go-compilers package.

**Comment by ingvagabund** [comment]
Resolved via 0c7ce13f5ed577371e9093033a400109aa17baa5.

For distributions where there is no go-srpm-macros, use 'gofed repo2spec' with --with-extra option. It will generate definition of %gobuild and %gotest macros as well.
